{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Quality control prior to EasyQC\n",
    "\n",
    "This is script to perform minimal quality control of the raw summary statistics which are then fed into EasyQC.\n",
    "\n",
    "It uses 1000 Genome phase 3 as reference.\n",
    "\n",
    "Missing values will be indicated with a period.\n",
    "\n",
    "The minimal QC includes following:\n",
    "- Changing GRCh build to 37\n",
    "- rsID from 1kG\n",
    "- uniqID from 1kG\n",
    "- Code missing data as NA\n",
    "- Calculation of MAF\n",
    "    - Consider effective allele as minor allele\n",
    "- Rename columns as below:\n",
    "    - CHR: Chromosome\n",
    "    - POS: base position (GCRh 37, hg19)\n",
    "    - SNP: rsID\n",
    "    - A1: effective allele\n",
    "    - A2: reference allele\n",
    "    - MAF: minor allele frequency\n",
    "    - Effect: beta or odds ratio\n",
    "    - SE: standard error of the effect\n",
    "    - Pval: p-value\n",
    "    - N: Leverage SNP-specific sample size if present, otherwise use total N or effective sample size (in the case of binary trait)\n",
    "    - INFO"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Utility functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from math import isnan\n",
    "\n",
    "def call_variant_matching_reference():\n",
    "    file = \"1kG_chrpos_rsID.tsv\"\n",
    "    df = pd.read_csv(file, sep='\\t', index_col=False)\n",
    "    return df\n",
    "\n",
    "def is_na(x):\n",
    "    if isinstance(x, float):\n",
    "        return True if isnan(x) else False\n",
    "    return True if x in ['NA', 'NAN', 'nan', 'None', 'none', None] else False\n",
    "\n",
    "\n",
    "def is_floatNan(x):\n",
    "    return True if isnan(x) else False\n",
    "\n",
    "\n",
    "def is_strna(x):\n",
    "    return True if x in ['NA', 'NAN', 'nan', 'None', 'none'] else False\n",
    "\n",
    "def is_integer(x):\n",
    "    try:\n",
    "        int(x)\n",
    "        return True\n",
    "    except:\n",
    "        return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "var_match_ref = call_variant_matching_reference() # Index(['CHRPOS', 'rsID'], dtype='object')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Body mass index (BMI)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Body mass index (BMI) / Cohort: GIANT 2018 / Pulit et al., 2018\n",
    "\n",
    "- File name: Bmi.giant-ukbb.meta-analysis.combined.23May2018.HapMap2_only.txt\n",
    "- Whitespace-separated\n",
    "- Columns: CHR POS SNP Tested_Allele Other_Allele Freq_Tested_Allele BETA SE P N INFO\n",
    "- GRCh37\n",
    "\n",
    "QC performed:\n",
    "- Code missing data as NA\n",
    "- SNP allele separation : A small number of SNPs (<9,000) from the GIANT data had no dbSNP151 identifier, and are left as just an rsID.\n",
    "- Missing CHR and POS from 1kG\n",
    "- MAF calculation and check\n",
    "- Rename columns\n",
    "- Reorder columns\n",
    "- Assign appropriate data type except NA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "file = \"Bmi.giant-ukbb.meta-analysis.combined.23May2018.HapMap2_only.txt\"\n",
    "df = pd.read_csv(file, sep=' ', index_col=False)\n",
    "\n",
    "# Code missing data as NA\n",
    "df['CHR'] = df['CHR'].apply(lambda x: 'NA' if is_floatNan(x) else str(int(x)))\n",
    "df['POS'] = df['POS'].apply(lambda x: 'NA' if is_floatNan(x) else str(int(x)))\n",
    "df['SNP'] = df['SNP'].apply(lambda x: 'NA' if is_strna(x) else str(x))\n",
    "df['Tested_Allele'] = df['Tested_Allele'].apply(lambda x: 'NA' if is_strna(x) else str(x))\n",
    "df['Other_Allele'] = df['Other_Allele'].apply(lambda x: 'NA' if is_strna(x) else str(x))\n",
    "df['Freq_Tested_Allele'] = df['Freq_Tested_Allele'].apply(lambda x: 'NA' if is_floatNan(x) else str(float(x)))\n",
    "df['BETA'] = df['BETA'].apply(lambda x: 'NA' if is_floatNan(x) else str(float(x)))\n",
    "df['SE'] = df['SE'].apply(lambda x: 'NA' if is_floatNan(x) else str(float(x)))\n",
    "df['P'] = df['P'].apply(lambda x: 'NA' if is_floatNan(x) else str(float(x)))\n",
    "df['N'] = df['N'].apply(lambda x: 'NA' if is_floatNan(x) else str(float(x)))\n",
    "df['INFO'] = df['INFO'].apply(lambda x: 'NA' if is_floatNan(x) else str(float(x)))\n",
    "\n",
    "# SNP allele separation\n",
    "df['rsID'] = df['SNP'].apply(lambda snp: snp.split(sep=':')[0] if snp != 'NA' else 'NA')\n",
    "df.drop(columns=['SNP'], inplace=True)\n",
    "\n",
    "# CHR and POS from 1kG\n",
    "df = df.merge(var_match_ref, left_on='rsID', right_on='rsID', how='left')\n",
    "\n",
    "def check_chrpos_na(x):\n",
    "    try:\n",
    "        if is_floatNan(x):\n",
    "            return True\n",
    "        return False\n",
    "    except:\n",
    "        if is_strna(x):\n",
    "            return True\n",
    "        return False\n",
    "\n",
    "df['CHRPOS'] = df['CHRPOS'].apply(lambda x: 'NA' if check_chrpos_na(x) else str(x))\n",
    "df['CHR'] = df['CHRPOS'].apply(lambda uniqid: 'NA' if uniqid == 'NA' else str(uniqid.split(sep=':')[0]))\n",
    "df['POS'] = df['CHRPOS'].apply(lambda uniqid: 'NA' if uniqid == 'NA' else str(uniqid.split(sep=':')[1]))\n",
    "df.drop(columns=['CHRPOS'])\n",
    "\n",
    "# MAF calculation and check\n",
    "df['MAF check'] = df['Freq_Tested_Allele'].apply(lambda eaf: 'NA' if eaf == 'NA' else True if float(eaf) <= 0.5 else False)\n",
    "df['MAF_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['Freq_Tested_Allele'] if row['MAF check'] else str(1 - float(row['Freq_Tested_Allele'])), axis=1)\n",
    "df['A1_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['Tested_Allele'] if row['MAF check'] else row['Other_Allele'], axis=1)\n",
    "df['A2_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['Other_Allele'] if row['MAF check'] else row['Tested_Allele'], axis=1)\n",
    "df['Effect'] =  df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['BETA'] if row['MAF check'] else str(-1 * float(row['BETA'])), axis=1)\n",
    "df.drop(columns=['Freq_Tested_Allele', 'MAF check', 'Tested_Allele', 'Other_Allele', 'BETA'], inplace=True)\n",
    "\n",
    "# Rename columns\n",
    "df.rename(columns={'se':'SE', 'P':'Pval', 'MAF_new':'MAF', 'A1_new':'A1', 'A2_new':'A2', 'rsID':'SNP'}, inplace=True)\n",
    "\n",
    "# Reorder columns\n",
    "df = df[['CHR', 'POS', 'SNP', 'A1', 'A2', 'MAF', 'Effect', 'SE', 'Pval', 'N', 'INFO']]\n",
    "\n",
    "# Assign appropriate data type except NA\n",
    "df['CHR'] = df['CHR'].apply(lambda x: 'NA' if x == 'NA' else int(x))\n",
    "df['POS'] = df['POS'].apply(lambda x: 'NA' if x == 'NA' else int(x))\n",
    "df['SNP'] = df['SNP'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['A1'] = df['A1'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['A2'] = df['A2'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['MAF'] = df['MAF'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['Effect'] = df['Effect'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['SE'] = df['SE'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['Pval'] = df['Pval'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['N'] = df['N'].apply(lambda x: 'NA' if x == 'NA' else int(float(x)))\n",
    "df['INFO'] = df['INFO'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "\n",
    "# Reorder according to CHR and POS\n",
    "df.sort_values(by=['CHR', 'POS'], ascending=True, inplace=True)\n",
    "\n",
    "# Convert 'NA' to .\n",
    "df.replace(to_replace=['NA'],\n",
    "            value='.',\n",
    "            inplace=True\n",
    "            )\n",
    "\n",
    "# Save\n",
    "save_filename = \"pQC.Bmi.giant-ukbb.meta-analysis.combined.23May2018.HapMap2_only.txt\"\n",
    "df.to_csv(save_filename, sep='\\t', index=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Body mass index (BMI) / Cohort: GIANT 2015 / Locke et al., 2015\n",
    "\n",
    "- File name: SNP_gwas_mc_merge_nogc.tbl.uniq\n",
    "- tab-separated\n",
    "- Columns: SNP     A1      A2      Freq1.Hapmap    b       se      p       N\n",
    "- GRCh37\n",
    "\n",
    "QC performed:\n",
    "- Code missing data as NA\n",
    "- CHR and POS from 1kG\n",
    "- MAF calculation and check\n",
    "- Rename columns\n",
    "- Reorder columns\n",
    "- Assign appropriate data type except NA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "file = \"SNP_gwas_mc_merge_nogc.tbl.uniq\"\n",
    "df = pd.read_csv(file, sep='\\t', index_col=False)\n",
    "\n",
    "# Code missing data as NA\n",
    "df['SNP'] = df['SNP'].apply(lambda x: 'NA' if is_strna(x) else str(x))\n",
    "df['A1'] = df['A1'].apply(lambda x: 'NA' if is_strna(x) else str(x))\n",
    "df['A2'] = df['A2'].apply(lambda x: 'NA' if is_strna(x) else str(x))\n",
    "df['Freq1.Hapmap'] = df['Freq1.Hapmap'].apply(lambda x: 'NA' if is_floatNan(x) else str(float(x)))\n",
    "df['b'] = df['b'].apply(lambda x: 'NA' if is_floatNan(x) else str(float(x)))\n",
    "df['se'] = df['se'].apply(lambda x: 'NA' if is_floatNan(x) else str(float(x)))\n",
    "df['p'] = df['p'].apply(lambda x: 'NA' if is_floatNan(x) else str(float(x)))\n",
    "df['N'] = df['N'].apply(lambda x: 'NA' if is_floatNan(x) else str(float(x)))\n",
    "\n",
    "# CHR and POS from 1kG\n",
    "df = df.merge(var_match_ref, left_on='SNP', right_on='rsID', how='left')\n",
    "\n",
    "def check_chrpos_na(x):\n",
    "    try:\n",
    "        if is_floatNan(x):\n",
    "            return True\n",
    "        return False\n",
    "    except:\n",
    "        if is_strna(x):\n",
    "            return True\n",
    "        return False\n",
    "\n",
    "df['CHRPOS'] = df['CHRPOS'].apply(lambda x: 'NA' if check_chrpos_na(x) else str(x))\n",
    "df['CHR'] = df['CHRPOS'].apply(lambda uniqid: 'NA' if uniqid == 'NA' else str(uniqid.split(sep=':')[0]))\n",
    "df['POS'] = df['CHRPOS'].apply(lambda uniqid: 'NA' if uniqid == 'NA' else str(uniqid.split(sep=':')[1]))\n",
    "df.drop(columns=['CHRPOS', 'rsID'])\n",
    "\n",
    "# MAF calculation and check\n",
    "df['MAF check'] = df['Freq1.Hapmap'].apply(lambda eaf: 'NA' if eaf == 'NA' else True if float(eaf) <= 0.5 else False)\n",
    "df['MAF_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['Freq1.Hapmap'] if row['MAF check'] else str(1 - float(row['Freq1.Hapmap'])), axis=1)\n",
    "df['A1_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['A1'] if row['MAF check'] else row['A2'], axis=1)\n",
    "df['A2_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['A2'] if row['MAF check'] else row['A1'], axis=1)\n",
    "df['Effect'] =  df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['b'] if row['MAF check'] else str(-1 * float(row['b'])), axis=1)\n",
    "df.drop(columns=['Freq1.Hapmap', 'MAF check', 'A1', 'A2', 'b'], inplace=True)\n",
    "\n",
    "# Rename columns\n",
    "df.rename(columns={'se':'SE', 'p':'Pval', 'MAF_new':'MAF', 'A1_new':'A1', 'A2_new':'A2'}, inplace=True)\n",
    "\n",
    "# Reorder columns\n",
    "df = df[['CHR', 'POS', 'SNP', 'A1', 'A2', 'MAF', 'Effect', 'SE', 'Pval', 'N']]\n",
    "\n",
    "# Assign appropriate data type except NA\n",
    "df['CHR'] = df['CHR'].apply(lambda x: 'NA' if x == 'NA' else int(x))\n",
    "df['POS'] = df['POS'].apply(lambda x: 'NA' if x == 'NA' else int(x))\n",
    "df['SNP'] = df['SNP'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['A1'] = df['A1'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['A2'] = df['A2'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['MAF'] = df['MAF'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['Effect'] = df['Effect'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['SE'] = df['SE'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['Pval'] = df['Pval'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['N'] = df['N'].apply(lambda x: 'NA' if x == 'NA' else int(float(x)))\n",
    "\n",
    "# Reorder according to CHR and POS\n",
    "df.sort_values(by=['CHR', 'POS'], ascending=True, inplace=True)\n",
    "\n",
    "\n",
    "# Re-code 'NA' to .\n",
    "df.replace(to_replace=['NA'],\n",
    "            value='.',\n",
    "            inplace=True\n",
    "            )\n",
    "\n",
    "# Save\n",
    "save_filename =\"pQC.SNP_gwas_mc_merge_nogc.tbl.uniq.txt\"\n",
    "df.to_csv(save_filename, sep='\\t', index=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Fasting glucose (FG)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fasting glucose (FG) / Cohort: MAGIC / Lagou et al., 2021\n",
    "\n",
    "- File name: FG_combined_1000G_density_formatted_21-03-29.txt\n",
    "- tab-separated\n",
    "- Columns: z       source  rsid    a1      a2      r2.pred p-value n       maf     beta    se\n",
    "- GRCh37\n",
    "- A1 is the effect allele\n",
    "\n",
    "README:\\\n",
    "a. z – Z-score of association for FG or FI;\\\n",
    "b. source – SSIMP for imputed and GWAS for SNPs present in the HapMap density;\\\n",
    "c. rsid – rsID;\\\n",
    "d. a1 – reference allele;\\\n",
    "e. a2 – effect allele;\\\n",
    "f. r2.pred – SS-imp imputation quality;\\\n",
    "g. p-value – P-value of association;\\\n",
    "h. n – sample size;\\\n",
    "i. maf – minor allele frequency;\\\n",
    "j. beta – effect size for FG or FI;\\\n",
    "k. se - standard error of the effect estimate for FG or FI.\\\n",
    "\n",
    "QC performed:\n",
    "- Remove unneccesary column: source\n",
    "- Code missing data as NA\n",
    "- Rename a1 as A2, a2 as A1\n",
    "- CHR and POS from 1kG\n",
    "- MAF check\n",
    "- Rename columns\n",
    "- Reorder columns\n",
    "- Assign appropriate data type except NA\n",
    "- Re-code 'NA' to ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "file = \"FG_combined_1000G_density_formatted_21-03-29.txt\"\n",
    "df = pd.read_csv(file, sep='\\t', index_col=False)\n",
    "\n",
    "# Remove unneccesary column: source\n",
    "df.drop(columns=['source'], inplace=True)\n",
    "\n",
    "# Code missing data as NA\n",
    "df['z'] = df['z'].apply(lambda x: 'NA' if is_floatNan(x) else str(float(x)))\n",
    "df['rsid'] = df['rsid'].apply(lambda x: 'NA' if is_strna(x) else str(x))\n",
    "df['a1'] = df['a1'].apply(lambda x: 'NA' if is_strna(x) else str(x))\n",
    "df['a2'] = df['a2'].apply(lambda x: 'NA' if is_strna(x) else str(x))\n",
    "df['maf'] = df['maf'].apply(lambda x: 'NA' if is_floatNan(x) else str(float(x)))\n",
    "df['beta'] = df['beta'].apply(lambda x: 'NA' if is_floatNan(x) else str(float(x)))\n",
    "df['se'] = df['se'].apply(lambda x: 'NA' if is_floatNan(x) else str(float(x)))\n",
    "df['p-value'] = df['p-value'].apply(lambda x: 'NA' if is_floatNan(x) else str(float(x)))\n",
    "df['n'] = df['n'].apply(lambda x: 'NA' if is_floatNan(x) else str(float(x)))\n",
    "df['r2.pred'] = df['r2.pred'].apply(lambda x: 'NA' if is_floatNan(x) else str(float(x)))\n",
    "\n",
    "# Rename a1 as A2, a2 as A1\n",
    "df.rename(columns={'a1':'A1', 'a2':'A2'}, inplace=True)\n",
    "\n",
    "# CHR and POS from 1kG\n",
    "df = df.merge(var_match_ref, left_on='rsid', right_on='rsID', how='left')\n",
    "\n",
    "def check_chrpos_na(x):\n",
    "    try:\n",
    "        if is_floatNan(x):\n",
    "            return True\n",
    "        return False\n",
    "    except:\n",
    "        if is_strna(x):\n",
    "            return True\n",
    "        return False\n",
    "\n",
    "df['CHRPOS'] = df['CHRPOS'].apply(lambda x: 'NA' if check_chrpos_na(x) else str(x))\n",
    "df['CHR'] = df['CHRPOS'].apply(lambda uniqid: 'NA' if uniqid == 'NA' else str(uniqid.split(sep=':')[0]))\n",
    "df['POS'] = df['CHRPOS'].apply(lambda uniqid: 'NA' if uniqid == 'NA' else str(uniqid.split(sep=':')[1]))\n",
    "df.drop(columns=['CHRPOS', 'rsID'], inplace=True)\n",
    "\n",
    "# MAF calculation and check\n",
    "df['MAF check'] = df['maf'].apply(lambda x: 'NA' if x == 'NA' else True if float(x) <= 0.5 else False)\n",
    "df['MAF_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['maf'] if row['MAF check'] else str(1 - float(row['maf'])), axis=1)\n",
    "df['A1_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['A1'] if row['MAF check'] else row['A2'], axis=1)\n",
    "df['A2_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['A2'] if row['MAF check'] else row['A1'], axis=1)\n",
    "df['Effect'] =  df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['beta'] if row['MAF check'] else str(-1 * float(row['beta'])), axis=1)\n",
    "df.drop(columns=['maf', 'MAF check', 'A1', 'A2', 'beta'], inplace=True)\n",
    "\n",
    "# Rename columns\n",
    "df.rename(columns={'rsid':'SNP', 'r2.pred':'INFO', 'p-value':'Pval', 'n':'N', 'se':'SE', 'MAF_new':'MAF', 'A1_new':'A1', 'A2_new':'A2'}, inplace=True)\n",
    "\n",
    "# Reorder columns\n",
    "df = df[['CHR', 'POS', 'SNP', 'A1', 'A2', 'MAF', 'Effect', 'SE', 'Pval', 'N', 'INFO']]\n",
    "\n",
    "# Assign appropriate data type except NA\n",
    "df['CHR'] = df['CHR'].apply(lambda x: 'NA' if x == 'NA' else int(x))\n",
    "df['POS'] = df['POS'].apply(lambda x: 'NA' if x == 'NA' else int(x))\n",
    "df['SNP'] = df['SNP'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['A1'] = df['A1'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['A2'] = df['A2'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['MAF'] = df['MAF'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['Effect'] = df['Effect'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['SE'] = df['SE'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['Pval'] = df['Pval'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['N'] = df['N'].apply(lambda x: 'NA' if x == 'NA' else int(float(x)))\n",
    "df['INFO'] = df['INFO'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "\n",
    "# Reorder according to CHR and POS\n",
    "df.sort_values(by=['CHR', 'POS'], ascending=True, inplace=True)\n",
    "\n",
    "\n",
    "# Re-code 'NA' to .\n",
    "df.replace(to_replace=['NA'],\n",
    "            value='.',\n",
    "            inplace=True\n",
    "            )\n",
    "\n",
    "# Save\n",
    "save_filename = \"pQC.FG_combined_1000G_density_formatted_21-03-29.A1asEffect.txt\"\n",
    "df.to_csv(save_filename, sep='\\t', index=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# High-density lipoprotein (HDL)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## High-density lipoprotein (HDL) / Cohort: GLGC 2022 with UKB\n",
    "\n",
    "- File name: HDL_INV_EUR_HRC_1KGP3_others_ALL.meta.singlevar.results.txt\n",
    "- tab-separated\n",
    "- Columns: rsID    CHROM   POS_b37 REF     ALT     N       N_studies       POOLED_ALT_AF   EFFECT_SIZE     SE      pvalue_neg_log10        pvalue  pvalue_neg_log10_GC     pvalue_GC\n",
    "- GRCh37\n",
    "\n",
    "Columns:\\\n",
    "rsID: ID from dbSNP 150\\\n",
    "CHROM: Chromosome\\\n",
    "POS_b37: Position in build 37\\\n",
    "REF: non-effect allele\\\n",
    "ALT: effect allele\\\n",
    "N: total number of individuals with results for this variant\\\n",
    "N_studies: total number of studies with results for this variant\\\n",
    "POOLED_ALT_AF: allele frequency of the ALT allele\\\n",
    "EFFECT_SIZE: effect size of the ALT allele\\\n",
    "SE: standard error\\\n",
    "pvalue_neg_log10: -log10(p-value)\\\n",
    "pvalue: p-value\\\n",
    "pvalue_neg_log10_GC: -log10(GC-corrected p-value)\\\n",
    "pvalue_GC: GC-corrected p-value\\\n",
    "\n",
    "QC performed:\n",
    "- Remove unneccesary column: N_studies, pvalue_neg_log10, pvalue_neg_log10_GC, pvalue_GC\n",
    "- Code missing data as NA\n",
    "- MAF check\n",
    "- Rename columns\n",
    "- Reorder columns\n",
    "- Assign appropriate data type except NA\n",
    "- Re-code 'NA' to ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "file = \"HDL_INV_EUR_HRC_1KGP3_others_ALL.meta.singlevar.results.txt\"\n",
    "df = pd.read_csv(file, sep='\\t', index_col=False)\n",
    "\n",
    "# Remove unneccesary columns\n",
    "df.drop(columns=['N_studies', 'pvalue_neg_log10', 'pvalue_neg_log10_GC', 'pvalue_GC'], inplace=True)\n",
    "\n",
    "# Code missing data as NA\n",
    "df['rsID'] = df['rsID'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['CHROM'] = df['CHROM'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['POS_b37'] = df['POS_b37'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['REF'] = df['REF'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['ALT'] = df['ALT'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['POOLED_ALT_AF'] = df['POOLED_ALT_AF'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['EFFECT_SIZE'] = df['EFFECT_SIZE'].apply(lambda x: 'NA' if is_na(x) else str(-1*float(x)))\n",
    "df['SE'] = df['SE'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['pvalue'] = df['pvalue'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['N'] = df['N'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "\n",
    "# MAF calculation and check\n",
    "df['MAF check'] = df['POOLED_ALT_AF'].apply(lambda x: 'NA' if x == 'NA' else True if float(x) <= 0.5 else False)\n",
    "df['MAF_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['POOLED_ALT_AF'] if row['MAF check'] else str(1 - float(row['POOLED_ALT_AF'])), axis=1)\n",
    "df['A1_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['ALT'] if row['MAF check'] else row['REF'], axis=1)\n",
    "df['A2_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['REF'] if row['MAF check'] else row['ALT'], axis=1)\n",
    "df['Effect'] =  df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['EFFECT_SIZE'] if row['MAF check'] else str(-1 * float(row['EFFECT_SIZE'])), axis=1)\n",
    "df.drop(columns=['POOLED_ALT_AF', 'MAF check', 'ALT', 'REF', 'EFFECT_SIZE'], inplace=True)\n",
    "\n",
    "# Rename columns\n",
    "df.rename(columns={'rsID':'SNP', 'CHROM':'CHR', 'POS_b37':'POS', 'A2_new':'A2', 'A1_new':'A1', 'MAF_new':'MAF', 'pvalue':'Pval'}, inplace=True)\n",
    "\n",
    "# Reorder columns\n",
    "df = df[['CHR', 'POS', 'SNP', 'A1', 'A2', 'MAF', 'Effect', 'SE', 'Pval', 'N']]\n",
    "\n",
    "# Assign appropriate data type except NA\n",
    "df['CHR'] = df['CHR'].apply(lambda x: 'NA' if x == 'NA' else int(x))\n",
    "df['POS'] = df['POS'].apply(lambda x: 'NA' if x == 'NA' else int(x))\n",
    "df['SNP'] = df['SNP'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['A1'] = df['A1'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['A2'] = df['A2'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['MAF'] = df['MAF'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['Effect'] = df['Effect'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['SE'] = df['SE'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['Pval'] = df['Pval'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['N'] = df['N'].apply(lambda x: 'NA' if x == 'NA' else int(float(x)))\n",
    "\n",
    "# Reorder according to CHR and POS\n",
    "df.sort_values(by=['CHR', 'POS'], ascending=True, inplace=True)\n",
    "\n",
    "\n",
    "# Re-code 'NA' to .\n",
    "df.replace(to_replace=['NA'],\n",
    "            value='.',\n",
    "            inplace=True\n",
    "            )\n",
    "\n",
    "# Save\n",
    "save_filename = \"pQC.HDL_INV_EUR_HRC_1KGP3_others_ALL.meta.singlevar.results.txt\"\n",
    "df.to_csv(save_filename, sep='\\t', index=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## High-density lipoprotein (HDL) / Cohort: GLGC 2022 without UKB\n",
    "\n",
    "- File name: HDL_INV_EUR_HRC_1KGP3_others_ALL.meta.singlevar.results.txt\n",
    "- tab-separated\n",
    "- Columns: rsID    CHROM   POS_b37 REF     ALT     N       N_studies       POOLED_ALT_AF   EFFECT_SIZE     SE      pvalue_neg_log10        pvalue  pvalue_neg_log10_GC     pvalue_GC\n",
    "- GRCh37\n",
    "\n",
    "Columns:\\\n",
    "rsID: ID from dbSNP 150\\\n",
    "CHROM: Chromosome\\\n",
    "POS_b37: Position in build 37\\\n",
    "REF: non-effect allele\\\n",
    "ALT: effect allele\\\n",
    "N: total number of individuals with results for this variant\\\n",
    "N_studies: total number of studies with results for this variant\\\n",
    "POOLED_ALT_AF: allele frequency of the ALT allele\\\n",
    "EFFECT_SIZE: effect size of the ALT allele\\\n",
    "SE: standard error\\\n",
    "pvalue_neg_log10: -log10(p-value)\\\n",
    "pvalue: p-value\\\n",
    "pvalue_neg_log10_GC: -log10(GC-corrected p-value)\\\n",
    "pvalue_GC: GC-corrected p-value\\\n",
    "\n",
    "QC performed:\n",
    "- Remove unneccesary column: N_studies, pvalue_neg_log10, pvalue_neg_log10_GC, pvalue_GC\n",
    "- Code missing data as NA\n",
    "- MAF check\n",
    "- Rename columns\n",
    "- Reorder columns\n",
    "- Assign appropriate data type except NA\n",
    "- Re-code 'NA' to ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "file = \"without_UKB_HDL_INV_EUR_HRC_1KGP3_others_ALL.meta.singlevar.results\"\n",
    "df = pd.read_csv(file, sep='\\t', index_col=False)\n",
    "\n",
    "# Remove unneccesary columns\n",
    "df.drop(columns=['N_studies', 'pvalue_neg_log10', 'pvalue_neg_log10_GC', 'pvalue_GC'], inplace=True)\n",
    "\n",
    "# Code missing data as NA\n",
    "df['rsID'] = df['rsID'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['CHROM'] = df['CHROM'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['POS_b37'] = df['POS_b37'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['REF'] = df['REF'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['ALT'] = df['ALT'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['POOLED_ALT_AF'] = df['POOLED_ALT_AF'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['EFFECT_SIZE'] = df['EFFECT_SIZE'].apply(lambda x: 'NA' if is_na(x) else str(-1*float(x)))\n",
    "df['SE'] = df['SE'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['pvalue'] = df['pvalue'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['N'] = df['N'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "\n",
    "# MAF calculation and check\n",
    "df['MAF check'] = df['POOLED_ALT_AF'].apply(lambda x: 'NA' if x == 'NA' else True if float(x) <= 0.5 else False)\n",
    "df['MAF_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['POOLED_ALT_AF'] if row['MAF check'] else str(1 - float(row['POOLED_ALT_AF'])), axis=1)\n",
    "df['A1_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['ALT'] if row['MAF check'] else row['REF'], axis=1)\n",
    "df['A2_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['REF'] if row['MAF check'] else row['ALT'], axis=1)\n",
    "df['Effect'] =  df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['EFFECT_SIZE'] if row['MAF check'] else str(-1 * float(row['EFFECT_SIZE'])), axis=1)\n",
    "df.drop(columns=['POOLED_ALT_AF', 'MAF check', 'ALT', 'REF', 'EFFECT_SIZE'], inplace=True)\n",
    "\n",
    "# Rename columns\n",
    "df.rename(columns={'rsID':'SNP', 'CHROM':'CHR', 'POS_b37':'POS', 'A2_new':'A2', 'A1_new':'A1', 'MAF_new':'MAF', 'pvalue':'Pval'}, inplace=True)\n",
    "\n",
    "# Reorder columns\n",
    "df = df[['CHR', 'POS', 'SNP', 'A1', 'A2', 'MAF', 'Effect', 'SE', 'Pval', 'N']]\n",
    "\n",
    "# Assign appropriate data type except NA\n",
    "df['CHR'] = df['CHR'].apply(lambda x: 'NA' if x == 'NA' else int(x))\n",
    "df['POS'] = df['POS'].apply(lambda x: 'NA' if x == 'NA' else int(x))\n",
    "df['SNP'] = df['SNP'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['A1'] = df['A1'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['A2'] = df['A2'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['MAF'] = df['MAF'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['Effect'] = df['Effect'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['SE'] = df['SE'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['Pval'] = df['Pval'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['N'] = df['N'].apply(lambda x: 'NA' if x == 'NA' else int(float(x)))\n",
    "\n",
    "# Reorder according to CHR and POS\n",
    "df.sort_values(by=['CHR', 'POS'], ascending=True, inplace=True)\n",
    "\n",
    "\n",
    "# Re-code 'NA' to .\n",
    "df.replace(to_replace=['NA'],\n",
    "            value='.',\n",
    "            inplace=True\n",
    "            )\n",
    "\n",
    "# Save\n",
    "save_filename = \"pQC.without_UKB_HDL_INV_EUR_HRC_1KGP3_others_ALL.meta.singlevar.results.txt\"\n",
    "df.to_csv(save_filename, sep='\\t', index=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Triglycerides (TG)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Triglycerides (TG) / Cohort: GLGC 2022 with UKB\n",
    "\n",
    "- File name: logTG_INV_EUR_HRC_1KGP3_others_ALL.meta.singlevar.results.txt\n",
    "- tab-separated\n",
    "- Columns: rsID    CHROM   POS_b37 REF     ALT     N       N_studies       POOLED_ALT_AF   EFFECT_SIZE     SE      pvalue_neg_log10        pvalue  pvalue_neg_log10_GC     pvalue_GC\n",
    "- GRCh37\n",
    "\n",
    "Columns:\\\n",
    "rsID: ID from dbSNP 150\\\n",
    "CHROM: Chromosome\\\n",
    "POS_b37: Position in build 37\\\n",
    "REF: non-effect allele\\\n",
    "ALT: effect allele\\\n",
    "N: total number of individuals with results for this variant\\\n",
    "N_studies: total number of studies with results for this variant\\\n",
    "POOLED_ALT_AF: allele frequency of the ALT allele\\\n",
    "EFFECT_SIZE: effect size of the ALT allele\\\n",
    "SE: standard error\\\n",
    "pvalue_neg_log10: -log10(p-value)\\\n",
    "pvalue: p-value\\\n",
    "pvalue_neg_log10_GC: -log10(GC-corrected p-value)\\\n",
    "pvalue_GC: GC-corrected p-value\\\n",
    "\n",
    "QC performed:\n",
    "- Remove unneccesary column: N_studies, pvalue_neg_log10, pvalue_neg_log10_GC, pvalue_GC\n",
    "- Code missing data as NA\n",
    "- MAF check\n",
    "- Rename columns\n",
    "- Reorder columns\n",
    "- Assign appropriate data type except NA\n",
    "- Re-code 'NA' to ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "file = \"logTG_INV_EUR_HRC_1KGP3_others_ALL.meta.singlevar.results.txt\"\n",
    "df = pd.read_csv(file, sep='\\t', index_col=False)\n",
    "\n",
    "# Remove unneccesary columns\n",
    "df.drop(columns=['N_studies', 'pvalue_neg_log10', 'pvalue_neg_log10_GC', 'pvalue_GC'], inplace=True)\n",
    "\n",
    "# Code missing data as NA\n",
    "df['rsID'] = df['rsID'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['CHROM'] = df['CHROM'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['POS_b37'] = df['POS_b37'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['REF'] = df['REF'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['ALT'] = df['ALT'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['POOLED_ALT_AF'] = df['POOLED_ALT_AF'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['EFFECT_SIZE'] = df['EFFECT_SIZE'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['SE'] = df['SE'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['pvalue'] = df['pvalue'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['N'] = df['N'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "\n",
    "# MAF calculation and check\n",
    "df['MAF check'] = df['POOLED_ALT_AF'].apply(lambda x: 'NA' if x == 'NA' else True if float(x) <= 0.5 else False)\n",
    "df['MAF_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['POOLED_ALT_AF'] if row['MAF check'] else str(1 - float(row['POOLED_ALT_AF'])), axis=1)\n",
    "df['A1_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['ALT'] if row['MAF check'] else row['REF'], axis=1)\n",
    "df['A2_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['REF'] if row['MAF check'] else row['ALT'], axis=1)\n",
    "df['Effect'] =  df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['EFFECT_SIZE'] if row['MAF check'] else str(-1 * float(row['EFFECT_SIZE'])), axis=1)\n",
    "df.drop(columns=['POOLED_ALT_AF', 'MAF check', 'ALT', 'REF', 'EFFECT_SIZE'], inplace=True)\n",
    "\n",
    "# Rename columns\n",
    "df.rename(columns={'rsID':'SNP', 'CHROM':'CHR', 'POS_b37':'POS', 'A2_new':'A2', 'A1_new':'A1', 'MAF_new':'MAF', 'pvalue':'Pval'}, inplace=True)\n",
    "\n",
    "# Reorder columns\n",
    "df = df[['CHR', 'POS', 'SNP', 'A1', 'A2', 'MAF', 'Effect', 'SE', 'Pval', 'N']]\n",
    "\n",
    "# Assign appropriate data type except NA\n",
    "df['CHR'] = df['CHR'].apply(lambda x: 'NA' if x == 'NA' else int(x))\n",
    "df['POS'] = df['POS'].apply(lambda x: 'NA' if x == 'NA' else int(x))\n",
    "df['SNP'] = df['SNP'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['A1'] = df['A1'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['A2'] = df['A2'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['MAF'] = df['MAF'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['Effect'] = df['Effect'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['SE'] = df['SE'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['Pval'] = df['Pval'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['N'] = df['N'].apply(lambda x: 'NA' if x == 'NA' else int(float(x)))\n",
    "\n",
    "# Reorder according to CHR and POS\n",
    "df.sort_values(by=['CHR', 'POS'], ascending=True, inplace=True)\n",
    "\n",
    "\n",
    "# Re-code 'NA' to .\n",
    "df.replace(to_replace=['NA'],\n",
    "            value='.',\n",
    "            inplace=True\n",
    "            )\n",
    "\n",
    "# Save\n",
    "save_filename = \"pQC.logTG_INV_EUR_HRC_1KGP3_others_ALL.meta.singlevar.results.txt\"\n",
    "df.to_csv(save_filename, sep='\\t', index=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Triglycerides (TG) / Cohort: GLGC 2022 without UKB\n",
    "\n",
    "- File name: without_UKB_logTG_INV_EUR_HRC_1KGP3_others_ALL.meta.singlevar.results\n",
    "- tab-separated\n",
    "- Columns: rsID    CHROM   POS_b37 REF     ALT     N       N_studies       POOLED_ALT_AF   EFFECT_SIZE     SE      pvalue_neg_log10        pvalue  pvalue_neg_log10_GC     pvalue_GC\n",
    "- GRCh37\n",
    "\n",
    "Columns:\\\n",
    "rsID: ID from dbSNP 150\\\n",
    "CHROM: Chromosome\\\n",
    "POS_b37: Position in build 37\\\n",
    "REF: non-effect allele\\\n",
    "ALT: effect allele\\\n",
    "N: total number of individuals with results for this variant\\\n",
    "N_studies: total number of studies with results for this variant\\\n",
    "POOLED_ALT_AF: allele frequency of the ALT allele\\\n",
    "EFFECT_SIZE: effect size of the ALT allele\\\n",
    "SE: standard error\\\n",
    "pvalue_neg_log10: -log10(p-value)\\\n",
    "pvalue: p-value\\\n",
    "pvalue_neg_log10_GC: -log10(GC-corrected p-value)\\\n",
    "pvalue_GC: GC-corrected p-value\\\n",
    "\n",
    "QC performed:\n",
    "- Remove unneccesary column: N_studies, pvalue_neg_log10, pvalue_neg_log10_GC, pvalue_GC\n",
    "- Code missing data as NA\n",
    "- MAF check\n",
    "- Rename columns\n",
    "- Reorder columns\n",
    "- Assign appropriate data type except NA\n",
    "- Re-code 'NA' to ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "file = \"without_UKB_logTG_INV_EUR_HRC_1KGP3_others_ALL.meta.singlevar.results\"\n",
    "df = pd.read_csv(file, sep='\\t', index_col=False)\n",
    "\n",
    "# Remove unneccesary columns\n",
    "df.drop(columns=['N_studies', 'pvalue_neg_log10', 'pvalue_neg_log10_GC', 'pvalue_GC'], inplace=True)\n",
    "\n",
    "# Code missing data as NA\n",
    "df['rsID'] = df['rsID'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['CHROM'] = df['CHROM'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['POS_b37'] = df['POS_b37'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['REF'] = df['REF'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['ALT'] = df['ALT'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['POOLED_ALT_AF'] = df['POOLED_ALT_AF'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['EFFECT_SIZE'] = df['EFFECT_SIZE'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['SE'] = df['SE'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['pvalue'] = df['pvalue'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['N'] = df['N'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "\n",
    "# MAF calculation and check\n",
    "df['MAF check'] = df['POOLED_ALT_AF'].apply(lambda x: 'NA' if x == 'NA' else True if float(x) <= 0.5 else False)\n",
    "df['MAF_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['POOLED_ALT_AF'] if row['MAF check'] else str(1 - float(row['POOLED_ALT_AF'])), axis=1)\n",
    "df['A1_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['ALT'] if row['MAF check'] else row['REF'], axis=1)\n",
    "df['A2_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['REF'] if row['MAF check'] else row['ALT'], axis=1)\n",
    "df['Effect'] =  df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['EFFECT_SIZE'] if row['MAF check'] else str(-1 * float(row['EFFECT_SIZE'])), axis=1)\n",
    "df.drop(columns=['POOLED_ALT_AF', 'MAF check', 'ALT', 'REF', 'EFFECT_SIZE'], inplace=True)\n",
    "\n",
    "# Rename columns\n",
    "df.rename(columns={'rsID':'SNP', 'CHROM':'CHR', 'POS_b37':'POS', 'A2_new':'A2', 'A1_new':'A1', 'MAF_new':'MAF', 'pvalue':'Pval'}, inplace=True)\n",
    "\n",
    "# Reorder columns\n",
    "df = df[['CHR', 'POS', 'SNP', 'A1', 'A2', 'MAF', 'Effect', 'SE', 'Pval', 'N']]\n",
    "\n",
    "# Assign appropriate data type except NA\n",
    "df['CHR'] = df['CHR'].apply(lambda x: 'NA' if x == 'NA' else int(x))\n",
    "df['POS'] = df['POS'].apply(lambda x: 'NA' if x == 'NA' else int(x))\n",
    "df['SNP'] = df['SNP'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['A1'] = df['A1'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['A2'] = df['A2'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['MAF'] = df['MAF'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['Effect'] = df['Effect'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['SE'] = df['SE'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['Pval'] = df['Pval'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['N'] = df['N'].apply(lambda x: 'NA' if x == 'NA' else int(float(x)))\n",
    "\n",
    "# Reorder according to CHR and POS\n",
    "df.sort_values(by=['CHR', 'POS'], ascending=True, inplace=True)\n",
    "\n",
    "\n",
    "# Re-code 'NA' to .\n",
    "df.replace(to_replace=['NA'],\n",
    "            value='.',\n",
    "            inplace=True\n",
    "            )\n",
    "\n",
    "# Save\n",
    "save_filename = \"pQC.without_UKB_logTG_INV_EUR_HRC_1KGP3_others_ALL.meta.singlevar.results.txt\"\n",
    "df.to_csv(save_filename, sep='\\t', index=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Hypertension (HTN)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Hypertension (HTN) / Cohort: Finngen release 7\n",
    "\n",
    "- File name: finngen_R7_I9_HYPERTENSION\n",
    "- tab-separated\n",
    "- Columns: \n",
    "- GRCh38\n",
    "\n",
    "README:\\\n",
    "#chrom: chromosome on build GRCh38 (1-23)\\\n",
    "pos: position in base pairs on build GRCh38\\\n",
    "ref: reference allele\\\n",
    "alt: alternative allele (effect allele)\\\n",
    "rsids: variant identifier\\\n",
    "nearest_genes: nearest gene(s) (comma separated) from variant\\\n",
    "pval: p-value from ​\\\n",
    "mlogp: -log10(p-value)\\\n",
    "beta: effect size (log(OR) scale) estimated with  for the alternative allele\\\n",
    "sebeta: standard error of effect size estimated with ​\\\n",
    "af_alt: alternative (effect) allele frequency\\\n",
    "af_alt_cases: alternative (effect) allele frequency among cases\\\n",
    "af_alt_controls: alternative (effect) allele frequency among controls\\\n",
    "\n",
    "QC performed:\n",
    "- liftover\n",
    "- Remove unneccesary column: mlogp, nearest_genes, af_alt_cases, af_alt_controls, n_hom_cases, n_hom_ref_cases, n_het_cases, n_hom_controls, n_hom_ref_controls, n_het_controls\n",
    "- Code missing data as NA\n",
    "- MAF check\n",
    "- Rename columns\n",
    "- Reorder columns\n",
    "- Assign appropriate data type except NA\n",
    "- Re-code 'NA' to ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Liftover\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "file = 'finngen_R7_I9_HYPERTENSION'\n",
    "lifting_dir = 'HTN_Finngen_r7'\n",
    "\n",
    "## Generate bed file\n",
    "df = pd.read_csv(file, sep='\\t', index_col=False)\n",
    "df_bed = df[['#chrom', 'pos', 'rsids']].copy()\n",
    "df_bed = df_bed[df_bed['rsids'].notnull()]\n",
    "df_bed['chr'] = df_bed['#chrom'].apply(lambda x: \"chr{}\".format(x))\n",
    "df_bed['bp-1'] = df_bed['pos'].apply(lambda x: int(x) - 1)\n",
    "df_bed = df_bed[['chr', 'bp-1', 'pos', 'rsids']]\n",
    "df_bed.to_csv(os.path.join(lifting_dir, \"HTN_Finngen_r7.bed\"), sep='\\t', header=False, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Perform liftover\n",
    "!liftOver HTN_Finngen_r7.bed GRCh38_to_GRCh37.chain.gz HTN_Finngen_r7.lifted HTN_Finngen_r7.unlifted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "from tqdm import tqdm\n",
    "import csv\n",
    "\n",
    "file = 'finngen_R7_I9_HYPERTENSION'\n",
    "df = pd.read_csv(file, sep='\\t', index_col=False)\n",
    "\n",
    "# Remove unneccesary column\n",
    "df.drop(columns=[\"mlogp\", \"nearest_genes\", \"af_alt_cases\", \"af_alt_controls\"], inplace=True)\n",
    "\n",
    "df_ = df[df['rsids'].notnull()].copy()\n",
    "df_rsid_na = df[~df['rsids'].notnull()].copy()\n",
    "\n",
    "## Match bp from liftover\n",
    "liftover_out = pd.read_csv(\"HTN_Finngen_r7.lifted\", sep='\\t', index_col=False, names=['CHR', 'BP-1', 'BP', 'SNP'])\n",
    "liftover_out.drop(columns=['BP-1'], inplace=True)\n",
    "\n",
    "## Add lifted SNPs in unlifted file which are having Duplicated in new tag\n",
    "unlifted_file = \"HTN_Finngen_r7.unlifted\"\n",
    "retain = []\n",
    "with open(unlifted_file, 'r') as f:\n",
    "    reader = csv.reader(f, delimiter='\\t')\n",
    "    tag = False\n",
    "    for row in reader:\n",
    "        if tag:\n",
    "            retain.append(row)\n",
    "            tag = False\n",
    "        if row[0] == '#Duplicated in new':\n",
    "            tag = True\n",
    "unlifted_retained = pd.DataFrame(retain, columns=['CHR', 'BP-1', 'BP', 'SNP'])\n",
    "unlifted_retained.drop(columns=['BP-1'], inplace=True)\n",
    "\n",
    "liftover_out = pd.concat([liftover_out, unlifted_retained])\n",
    "\n",
    "## Start matching\n",
    "lifted_df = None\n",
    "for chr_i in tqdm(range(1, 24), leave=False):\n",
    "    df_partial = df_.loc[df_['#chrom'] == chr_i]\n",
    "\n",
    "    lifted_df_partial = df_partial.merge(liftover_out, left_on='rsids', right_on='SNP', how='left')\n",
    "    lifted_df_partial.drop(columns=['#chrom', 'pos', 'rsids'], inplace=True)\n",
    "    lifted_df_partial['CHR'] = lifted_df_partial['CHR'].apply(lambda chr: int(chr) if is_integer(chr) else float('nan'))\n",
    "    lifted_df_partial['BP'] = lifted_df_partial['BP'].apply(lambda chr: int(chr) if is_integer(chr) else float('nan'))\n",
    "    lifted_df = pd.concat([lifted_df, lifted_df_partial])\n",
    "\n",
    "del liftover_out\n",
    "\n",
    "\n",
    "# Attach rsID missing df (these will be removed during EasyQC step)\n",
    "df_rsid_na.rename(columns={'#chrom':'CHR', 'pos':'BP', 'rsids':'SNP'}, inplace=True)\n",
    "df_rsid_na = df_rsid_na[['ref', 'alt', 'pval', 'beta', 'sebeta', 'af_alt', 'CHR', 'BP', 'SNP']]\n",
    "\n",
    "df = pd.concat([lifted_df, df_rsid_na])\n",
    "\n",
    "# Code missing data as NA\n",
    "df['CHR'] = df['CHR'].apply(lambda x: 'NA' if is_na(x) else str(int(x)))\n",
    "df['BP'] = df['BP'].apply(lambda x: 'NA' if is_na(x) else str(int(x)))\n",
    "df['SNP'] = df['SNP'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['ref'] = df['ref'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['alt'] = df['alt'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['pval'] = df['pval'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['beta'] = df['beta'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['sebeta'] = df['sebeta'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['af_alt'] = df['af_alt'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "\n",
    "\n",
    "# MAF check\n",
    "df['MAF check'] = df['af_alt'].apply(lambda x: 'NA' if x == 'NA' else True if float(x) <= 0.5 else False)\n",
    "df['MAF_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['af_alt'] if row['MAF check'] else str(1 - float(row['af_alt'])), axis=1)\n",
    "df['alt_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['alt'] if row['MAF check'] else row['ref'], axis=1)\n",
    "df['ref_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['ref'] if row['MAF check'] else row['alt'], axis=1)\n",
    "df['Effect'] =  df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['beta'] if row['MAF check'] else str(-1 * float(row['beta'])), axis=1)\n",
    "df.drop(columns=['af_alt', 'MAF check', 'alt', 'ref', 'beta'], inplace=True)\n",
    "\n",
    "# Rename columns\n",
    "df.rename(columns={'ref_new':'A2', 'alt_new':'A1', 'pval':'Pval', 'sebeta':'SE', 'BP':'POS', 'MAF_new':'MAF'}, inplace=True)\n",
    "\n",
    "# Reorder columns\n",
    "df = df[['CHR', 'POS', 'SNP', 'A1', 'A2', 'MAF', 'Effect', 'SE', 'Pval']]\n",
    "\n",
    "# Assign appropriate data type except NA\n",
    "df['CHR'] = df['CHR'].apply(lambda x: 'NA' if x == 'NA' else int(x))\n",
    "df['POS'] = df['POS'].apply(lambda x: 'NA' if x == 'NA' else int(x))\n",
    "df['SNP'] = df['SNP'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['A1'] = df['A1'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['A2'] = df['A2'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['MAF'] = df['MAF'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['Effect'] = df['Effect'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['SE'] = df['SE'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['Pval'] = df['Pval'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "\n",
    "# Reorder according to CHR and POS\n",
    "df.sort_values(by=['CHR', 'POS'], ascending=True, inplace=True)\n",
    "\n",
    "\n",
    "# Re-code 'NA' to .\n",
    "df.replace(to_replace=['NA'],\n",
    "            value='.',\n",
    "            inplace=True\n",
    "            )\n",
    "\n",
    "# Save\n",
    "save_filename = \"pQC.finngen_R7_I9_HYPERTENSION.txt\"\n",
    "df.to_csv(save_filename, sep='\\t', index=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Hypertension (HTN) / Cohort: UKB\n",
    "\n",
    "- File name: 20002_1065_logistic.EUR.sumstats.MACfilt.txt\n",
    "- tab-separated\n",
    "- Columns: SNP     CHR     BP      A1      TEST    NMISS   OR      SE      L95     U95     STAT    P       A2      MAF     NCHROBS SNPID_UKB       A1_UKB  A2_UKB  INFO_UKB        MAF_UKB\n",
    "- GRCh37\n",
    "\n",
    "README:\\\n",
    "SNP: unique ID of the SNP consists of chromosome, position and alphabetically ordered alleles\\\n",
    "CHR: chromosome\\\n",
    "BP: base pair position on GRCh37\\\n",
    "A1: effect allele\\\n",
    "TEST: Type of test (ADD for all files)\\\n",
    "NMISS: Number of non-missing genotypes\\\n",
    "BETA/OR: Regression coefficient or odds ratio\\\n",
    "SE: Standard error (for binary traits, SE in logOR scale)\\\n",
    "L95: Lower bound on confidence interval for CMH odds ratio\\\n",
    "U95: Upper bound on confidence interval for CMH odds ratio\\\n",
    "STAT: Coefficient t-statistics\\\n",
    "P: P-value\\\n",
    "A2: non effect allele\\\n",
    "MAF: Minor allele frequency\\\n",
    "NCHROBS: Number of allele observation\\\n",
    "SNPID_UKB: rsID provided by UK Biobank\\\n",
    "A1_UKB: A1 allele in UK Biobank\\\n",
    "A2_UKB: A2 allele in UK Biobank\\\n",
    "INFO_UKB: Info score provided by UK Biobank\\\n",
    "MAF_UKB: MAF of entire UK Boiobank samples\\\n",
    "\n",
    "QC performed:\n",
    "- Remove unneccesary column: TEST, NMISS, L95, U95, STAT, NCHROBS, A1_UKB, A2_UKB, MAF_UKB\n",
    "- Code missing data as NA\n",
    "- OR to beta\n",
    "- MAF check\n",
    "- Rename columns\n",
    "- Reorder columns\n",
    "- Assign appropriate data type except NA\n",
    "- Re-code 'NA' to ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "from math import log, exp\n",
    "\n",
    "file = \"20002_1065_logistic.EUR.sumstats.MACfilt.txt\"\n",
    "df = pd.read_csv(file, sep='\\t', index_col=False)\n",
    "\n",
    "# Remove unneccesary columns\n",
    "df.drop(columns=['SNP', 'TEST', 'NMISS', 'L95', 'U95', 'STAT', 'NCHROBS', 'A1_UKB', 'A2_UKB', 'MAF_UKB'], inplace=True)\n",
    "\n",
    "# Code missing data as NA\n",
    "df['SNPID_UKB'] = df['SNPID_UKB'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['CHR'] = df['CHR'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['BP'] = df['BP'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['A1'] = df['A1'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['A2'] = df['A2'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['MAF'] = df['MAF'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['OR'] = df['OR'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['SE'] = df['SE'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['P'] = df['P'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "\n",
    "# OR to ln(OR)\n",
    "df['Effect'] = df['OR'].apply(lambda x: 'NA' if x == 'NA' else str(log(float(x))))\n",
    "\n",
    "# MAF check\n",
    "df['MAF check'] = df['MAF'].apply(lambda x: 'NA' if x == 'NA' else True if float(x) <= 0.5 else False)\n",
    "df['MAF_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['MAF'] if row['MAF check'] else str(1 - float(row['MAF'])), axis=1)\n",
    "df['A1_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['A1'] if row['MAF check'] else row['A2'], axis=1)\n",
    "df['A2_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['A2'] if row['MAF check'] else row['A1'], axis=1)\n",
    "df['Effect_new'] =  df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['Effect'] if row['MAF check'] else str(-1 * float(row['Effect'])), axis=1)\n",
    "df['OR_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['OR'] if row['MAF check'] else str(exp(float(row['Effect_new']))), axis=1)\n",
    "df.drop(columns=['MAF', 'MAF check', 'A1', 'A2', 'Effect', 'OR'], inplace=True)\n",
    "\n",
    "\n",
    "# Rename columns\n",
    "df.rename(columns={'SNPID_UKB':'SNP', 'BP':'POS', 'P':'Pval', 'MAF_new':'MAF', 'A1_new':'A1', 'A2_new':'A2', 'Effect_new':'Effect', 'OR_new':'OR'}, inplace=True)\n",
    "\n",
    "# Reorder columns\n",
    "df = df[['CHR', 'POS', 'SNP', 'A1', 'A2', 'MAF', 'Effect', 'OR', 'SE', 'Pval']]\n",
    "\n",
    "# Assign appropriate data type except NA\n",
    "df['CHR'] = df['CHR'].apply(lambda x: 'NA' if x == 'NA' else int(x))\n",
    "df['POS'] = df['POS'].apply(lambda x: 'NA' if x == 'NA' else int(x))\n",
    "df['SNP'] = df['SNP'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['A1'] = df['A1'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['A2'] = df['A2'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['MAF'] = df['MAF'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['Effect'] = df['Effect'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['OR'] = df['OR'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['SE'] = df['SE'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['Pval'] = df['Pval'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "\n",
    "\n",
    "# Reorder according to CHR and POS\n",
    "df.sort_values(by=['CHR', 'POS'], ascending=True, inplace=True)\n",
    "\n",
    "\n",
    "# Re-code 'NA' to .\n",
    "df.replace(to_replace=['NA'],\n",
    "            value='.',\n",
    "            inplace=True\n",
    "            )\n",
    "\n",
    "# Save\n",
    "save_filename = \"pQC.20002_1065_logistic.EUR.sumstats.MACfilt.txt\"\n",
    "df.to_csv(save_filename, sep='\\t', index=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Hypertension (HTN) / Cohort: [METAL] FinnGen release 7, UKB\n",
    "\n",
    "- File name: ./meta_analysis/HTN_FinngenR7_UKB/meta.HTN.FinngenR7.UKBwatanabe.1.tbl\n",
    "- tab-separated\n",
    "- Columns: MarkerName      Allele1 Allele2 Freq1   FreqSE  Effect  StdErr  P-value Direction       HetISq  HetChiSq        HetDf   HetPVal\n",
    "- GRCh37\n",
    "\n",
    "README:\\\n",
    "\n",
    "\n",
    "QC performed:\n",
    "- Remove unneccesary column: FreqSE, Direction, HetISq, HetChiSq, HetDf, HetPVal\n",
    "- Code missing data as NA\n",
    "- CHR, POS matching from 1kG\n",
    "- MAF check\n",
    "- Rename columns\n",
    "- Reorder columns\n",
    "- Assign appropriate data type except NA\n",
    "- Re-code 'NA' to ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import pandas as pd\n",
    "import os\n",
    "from math import log, exp\n",
    "\n",
    "file = \"./meta_analysis/HTN_FinngenR7_UKB/meta.HTN.FinngenR7.UKBwatanabe.1.tbl\"\n",
    "df = pd.read_csv(file, sep='\\t', index_col=False)\n",
    "\n",
    "# Remove unneccesary columns\n",
    "df.drop(columns=['FreqSE', 'Direction', 'HetISq', 'HetChiSq', 'HetDf', 'HetPVal'], inplace=True)\n",
    "\n",
    "# Code missing data as NA\n",
    "df['MarkerName'] = df['MarkerName'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['Allele1'] = df['Allele1'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['Allele2'] = df['Allele2'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['Freq1'] = df['Freq1'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['Effect'] = df['Effect'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['StdErr'] = df['StdErr'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['P-value'] = df['P-value'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "\n",
    "# CHR and POS from 1kG\n",
    "df = df.merge(var_match_ref, left_on='MarkerName', right_on='rsID', how='left')\n",
    "\n",
    "def check_chrpos_na(x):\n",
    "    try:\n",
    "        if is_floatNan(x):\n",
    "            return True\n",
    "        return False\n",
    "    except:\n",
    "        if is_strna(x):\n",
    "            return True\n",
    "        return False\n",
    "\n",
    "df['CHRPOS'] = df['CHRPOS'].apply(lambda x: 'NA' if check_chrpos_na(x) else str(x))\n",
    "df['CHR'] = df['CHRPOS'].apply(lambda uniqid: 'NA' if uniqid == 'NA' else str(uniqid.split(sep=':')[0]))\n",
    "df['POS'] = df['CHRPOS'].apply(lambda uniqid: 'NA' if uniqid == 'NA' else str(uniqid.split(sep=':')[1]))\n",
    "df.drop(columns=['CHRPOS', 'rsID'], inplace=True)\n",
    "\n",
    "# MAF calculation and check\n",
    "df['MAF check'] = df['Freq1'].apply(lambda x: 'NA' if x == 'NA' else True if float(x) <= 0.5 else False)\n",
    "df['MAF_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['Freq1'] if row['MAF check'] else str(1 - float(row['Freq1'])), axis=1)\n",
    "df['A1_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['Allele1'] if row['MAF check'] else row['Allele2'], axis=1)\n",
    "df['A2_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['Allele2'] if row['MAF check'] else row['Allele1'], axis=1)\n",
    "df['Effect_new'] =  df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['Effect'] if row['MAF check'] else str(-1 * float(row['Effect'])), axis=1)\n",
    "df.drop(columns=['Freq1', 'MAF check', 'Allele1', 'Allele2', 'Effect'], inplace=True)\n",
    "\n",
    "# Rename columns\n",
    "df.rename(columns={'MarkerName':'SNP', 'A1_new':'A1', 'A2_new':'A2', 'Effect_new':'Effect', 'MAF_new':'MAF', 'StdErr':'SE', 'P-value':'Pval'}, inplace=True)\n",
    "\n",
    "# Reorder columns\n",
    "df = df[['CHR', 'POS', 'SNP', 'A1', 'A2', 'MAF', 'Effect', 'SE', 'Pval']]\n",
    "\n",
    "# Assign appropriate data type except NA\n",
    "df['CHR'] = df['CHR'].apply(lambda x: 'NA' if x == 'NA' else int(x))\n",
    "df['POS'] = df['POS'].apply(lambda x: 'NA' if x == 'NA' else int(x))\n",
    "df['SNP'] = df['SNP'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['A1'] = df['A1'].apply(lambda x: 'NA' if x == 'NA' else str(x).upper())\n",
    "df['A2'] = df['A2'].apply(lambda x: 'NA' if x == 'NA' else str(x).upper())\n",
    "df['MAF'] = df['MAF'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['Effect'] = df['Effect'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['SE'] = df['SE'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['Pval'] = df['Pval'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "\n",
    "# Reorder according to CHR and POS\n",
    "df.sort_values(by=['CHR', 'POS'], ascending=True, inplace=True)\n",
    "\n",
    "\n",
    "# Re-code 'NA' to .\n",
    "df.replace(to_replace=['NA'],\n",
    "            value='.',\n",
    "            inplace=True\n",
    "            )\n",
    "\n",
    "# Save\n",
    "save_filename = \"pQC.meta.HTN.FinngenR7.UKBwatanabe.1.tbl.txt\"\n",
    "df.to_csv(save_filename, sep='\\t', index=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Type II Diabetes (T2D)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Type II Diabetes (T2D) / Cohort: Finngen release 7\n",
    "\n",
    "- File name: finngen_R7_E4_DM2\n",
    "- tab-separated\n",
    "- Columns: \n",
    "- GRCh38\n",
    "\n",
    "README:\\\n",
    "#chrom: chromosome on build GRCh38 (1-23)\\\n",
    "pos: position in base pairs on build GRCh38\\\n",
    "ref: reference allele\\\n",
    "alt: alternative allele (effect allele)\\\n",
    "rsids: variant identifier\\\n",
    "nearest_genes: nearest gene(s) (comma separated) from variant\\\n",
    "pval: p-value from ​\\\n",
    "mlogp: -log10(p-value)\\\n",
    "beta: effect size (log(OR) scale) estimated with  for the alternative allele\\\n",
    "sebeta: standard error of effect size estimated with ​\\\n",
    "af_alt: alternative (effect) allele frequency\\\n",
    "af_alt_cases: alternative (effect) allele frequency among cases\\\n",
    "af_alt_controls: alternative (effect) allele frequency among controls\\\n",
    "\n",
    "QC performed:\n",
    "- liftover\n",
    "- Remove unneccesary column: mlogp, nearest_genes, af_alt_cases, af_alt_controls, n_hom_cases, n_hom_ref_cases, n_het_cases, n_hom_controls, n_hom_ref_controls, n_het_controls\n",
    "- Code missing data as NA\n",
    "- MAF check\n",
    "- Rename columns\n",
    "- Reorder columns\n",
    "- Assign appropriate data type except NA\n",
    "- Re-code 'NA' to ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Liftover\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "file = 'finngen_R7_E4_DM2'\n",
    "lifting_dir = './T2D_Finngen_r7'\n",
    "\n",
    "## Generate bed file\n",
    "df = pd.read_csv(file, sep='\\t', index_col=False)\n",
    "df_bed = df[['#chrom', 'pos', 'rsids']].copy()\n",
    "df_bed = df_bed[df_bed['rsids'].notnull()]\n",
    "df_bed['chr'] = df_bed['#chrom'].apply(lambda x: \"chr{}\".format(x))\n",
    "df_bed['bp-1'] = df_bed['pos'].apply(lambda x: int(x) - 1)\n",
    "df_bed = df_bed[['chr', 'bp-1', 'pos', 'rsids']]\n",
    "df_bed.to_csv(os.path.join(lifting_dir, \"T2D_Finngen_r7.bed\"), sep='\\t', header=False, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Perform liftover\n",
    "!liftOver T2D_Finngen_r7.bed GRCh38_to_GRCh37.chain.gz T2D_Finngen_r7.lifted T2D_Finngen_r7.unlifted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "from tqdm import tqdm\n",
    "import csv\n",
    "\n",
    "file = 'finngen_R7_E4_DM2'\n",
    "df = pd.read_csv(file, sep='\\t', index_col=False)\n",
    "\n",
    "# Remove unneccesary column\n",
    "df.drop(columns=[\"mlogp\", \"nearest_genes\", \"af_alt_cases\", \"af_alt_controls\"], inplace=True)\n",
    "\n",
    "df_ = df[df['rsids'].notnull()].copy()\n",
    "df_rsid_na = df[~df['rsids'].notnull()].copy()\n",
    "\n",
    "## Match bp from liftover\n",
    "liftover_out = pd.read_csv(\"/data1/sanghyeon/Projects/MetabolicSyndrome/GWASsumstat/litftover_sumstat/T2D_Finngen_r7/T2D_Finngen_r7.lifted\", sep='\\t', index_col=False, names=['CHR', 'BP-1', 'BP', 'SNP'])\n",
    "liftover_out.drop(columns=['BP-1'], inplace=True)\n",
    "\n",
    "## Add lifted SNPs in unlifted file which are having Duplicated in new tag\n",
    "unlifted_file = \"/data1/sanghyeon/Projects/MetabolicSyndrome/GWASsumstat/litftover_sumstat/T2D_Finngen_r7/T2D_Finngen_r7.unlifted\"\n",
    "retain = []\n",
    "with open(unlifted_file, 'r') as f:\n",
    "    reader = csv.reader(f, delimiter='\\t')\n",
    "    tag = False\n",
    "    for row in reader:\n",
    "        if tag:\n",
    "            retain.append(row)\n",
    "            tag = False\n",
    "        if row[0] == '#Duplicated in new':\n",
    "            tag = True\n",
    "unlifted_retained = pd.DataFrame(retain, columns=['CHR', 'BP-1', 'BP', 'SNP'])\n",
    "unlifted_retained.drop(columns=['BP-1'], inplace=True)\n",
    "\n",
    "liftover_out = pd.concat([liftover_out, unlifted_retained])\n",
    "\n",
    "## Start matching\n",
    "lifted_df = None\n",
    "for chr_i in tqdm(range(1, 24), leave=False):\n",
    "    df_partial = df_.loc[df_['#chrom'] == chr_i]\n",
    "\n",
    "    lifted_df_partial = df_partial.merge(liftover_out, left_on='rsids', right_on='SNP', how='left')\n",
    "    lifted_df_partial.drop(columns=['#chrom', 'pos', 'rsids'], inplace=True)\n",
    "    lifted_df_partial['CHR'] = lifted_df_partial['CHR'].apply(lambda chr: int(chr) if is_integer(chr) else float('nan'))\n",
    "    lifted_df_partial['BP'] = lifted_df_partial['BP'].apply(lambda chr: int(chr) if is_integer(chr) else float('nan'))\n",
    "    lifted_df = pd.concat([lifted_df, lifted_df_partial])\n",
    "\n",
    "del liftover_out\n",
    "\n",
    "\n",
    "# Attach rsID missing df (these will be removed during EasyQC step)\n",
    "df_rsid_na.rename(columns={'#chrom':'CHR', 'pos':'BP', 'rsids':'SNP'}, inplace=True)\n",
    "df_rsid_na = df_rsid_na[['ref', 'alt', 'pval', 'beta', 'sebeta', 'af_alt', 'CHR', 'BP', 'SNP']]\n",
    "\n",
    "df = pd.concat([lifted_df, df_rsid_na])\n",
    "\n",
    "# Code missing data as NA\n",
    "df['CHR'] = df['CHR'].apply(lambda x: 'NA' if is_na(x) else str(int(x)))\n",
    "df['BP'] = df['BP'].apply(lambda x: 'NA' if is_na(x) else str(int(x)))\n",
    "df['SNP'] = df['SNP'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['ref'] = df['ref'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['alt'] = df['alt'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['pval'] = df['pval'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['beta'] = df['beta'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['sebeta'] = df['sebeta'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['af_alt'] = df['af_alt'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "\n",
    "\n",
    "# MAF check\n",
    "df['MAF check'] = df['af_alt'].apply(lambda x: 'NA' if x == 'NA' else True if float(x) <= 0.5 else False)\n",
    "df['MAF_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['af_alt'] if row['MAF check'] else str(1 - float(row['af_alt'])), axis=1)\n",
    "df['alt_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['alt'] if row['MAF check'] else row['ref'], axis=1)\n",
    "df['ref_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['ref'] if row['MAF check'] else row['alt'], axis=1)\n",
    "df['Effect'] =  df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['beta'] if row['MAF check'] else str(-1 * float(row['beta'])), axis=1)\n",
    "df.drop(columns=['af_alt', 'MAF check', 'alt', 'ref', 'beta'], inplace=True)\n",
    "\n",
    "# Rename columns\n",
    "df.rename(columns={'ref_new':'A2', 'alt_new':'A1', 'pval':'Pval', 'sebeta':'SE', 'BP':'POS', 'MAF_new':'MAF'}, inplace=True)\n",
    "\n",
    "# Reorder columns\n",
    "df = df[['CHR', 'POS', 'SNP', 'A1', 'A2', 'MAF', 'Effect', 'SE', 'Pval']]\n",
    "\n",
    "# Assign appropriate data type except NA\n",
    "df['CHR'] = df['CHR'].apply(lambda x: 'NA' if x == 'NA' else int(x))\n",
    "df['POS'] = df['POS'].apply(lambda x: 'NA' if x == 'NA' else int(x))\n",
    "df['SNP'] = df['SNP'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['A1'] = df['A1'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['A2'] = df['A2'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['MAF'] = df['MAF'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['Effect'] = df['Effect'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['SE'] = df['SE'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['Pval'] = df['Pval'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "\n",
    "# Reorder according to CHR and POS\n",
    "df.sort_values(by=['CHR', 'POS'], ascending=True, inplace=True)\n",
    "\n",
    "\n",
    "# Re-code 'NA' to .\n",
    "df.replace(to_replace=['NA'],\n",
    "            value='.',\n",
    "            inplace=True\n",
    "            )\n",
    "\n",
    "# Save\n",
    "save_filename = \"pQC.finngen_R7_E4_DM2.txt\"\n",
    "df.to_csv(save_filename, sep='\\t', index=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Type II Diabetes (T2D) / Cohort: Mahajan 2022\n",
    "\n",
    "- File name: DIAMANTE-EUR.sumstat.txt\n",
    "- Whitespace-separated\n",
    "- Columns: chromosome(b37) position(b37) chrposID rsID effect_allele other_allele effect_allele_frequency Fixed-effects_beta Fixed-effects_SE Fixed-effects_p-value\n",
    "- GRCh37\n",
    "\n",
    "README:\\\n",
    "A space delimited file with one row per SNV including the following columns:\\\n",
    "chromosome and position (hg19, build 37); \\\n",
    "chromosome and position ID; \\\n",
    "rsID; \\\n",
    "effect allele and other allele; \\\n",
    "effect allele frequency; \\\n",
    "fixed-effects\n",
    "meta-analysis beta (log-odds ratio); \\\n",
    "fixed-effects meta-analysis standard error (corrected for genomic control); \\\n",
    "fixed-effects meta-analysis association p-value (corrected for genomic control).\n",
    "\n",
    "QC performed:\n",
    "- Drop unneccessary columns: chrposID\n",
    "- Rename columns\n",
    "- Code missing data as NA\n",
    "- Capitalize allele\n",
    "- MAF check\n",
    "- Rename columns\n",
    "- Reorder columns\n",
    "- Assign appropriate data type except NA\n",
    "- Re-code 'NA' to ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "file = 'DIAMANTE-EUR.sumstat.txt'\n",
    "df = pd.read_csv(file, sep=' ', index_col=False)\n",
    "\n",
    "# Drop unneccessary columns: chrposID\n",
    "df.drop(columns=['chrposID'], inplace=True)\n",
    "\n",
    "# Rename columns\n",
    "df.rename(columns={'chromosome(b37)':'CHR', 'position(b37)':'POS', 'rsID':'SNP', 'effect_allele':'A1', 'other_allele':'A2',\n",
    "                'effect_allele_frequency':'EAF', 'Fixed-effects_beta':'Effect', 'Fixed-effects_SE':'SE', 'Fixed-effects_p-value':'Pval'}, inplace=True)\n",
    "\n",
    "# Code missing data as NA\n",
    "df['CHR'] = df['CHR'].apply(lambda x: 'NA' if is_na(x) else str(int(x)))\n",
    "df['POS'] = df['POS'].apply(lambda x: 'NA' if is_na(x) else str(int(x)))\n",
    "df['SNP'] = df['SNP'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['A1'] = df['A1'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['A2'] = df['A2'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['Pval'] = df['Pval'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['Effect'] = df['Effect'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['SE'] = df['SE'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['EAF'] = df['EAF'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "\n",
    "# Capitalize allele\n",
    "df['A1'] = df['A1'].apply(lambda x: x.upper())\n",
    "df['A2'] = df['A2'].apply(lambda x: x.upper())\n",
    "\n",
    "# MAF check\n",
    "df['MAF check'] = df['EAF'].apply(lambda x: 'NA' if x == 'NA' else True if float(x) <= 0.5 else False)\n",
    "df['MAF'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['EAF'] if row['MAF check'] else str(1 - float(row['EAF'])), axis=1)\n",
    "df['A1_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['A1'] if row['MAF check'] else row['A2'], axis=1)\n",
    "df['A2_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['A2'] if row['MAF check'] else row['A1'], axis=1)\n",
    "df['Effect_new'] =  df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['Effect'] if row['MAF check'] else str(-1 * float(row['Effect'])), axis=1)\n",
    "df.drop(columns=['EAF', 'MAF check', 'A1', 'A2', 'Effect'], inplace=True)\n",
    "\n",
    "df.rename(columns={'A1_new':'A1', 'A2_new':'A2', 'Effect_new':'Effect'}, inplace=True)\n",
    "\n",
    "\n",
    "# Reorder columns\n",
    "df = df[['CHR', 'POS', 'SNP', 'A1', 'A2', 'MAF', 'Effect', 'SE', 'Pval']]\n",
    "\n",
    "# Assign appropriate data type except NA\n",
    "df['CHR'] = df['CHR'].apply(lambda x: 'NA' if x == 'NA' else int(x))\n",
    "df['POS'] = df['POS'].apply(lambda x: 'NA' if x == 'NA' else int(x))\n",
    "df['SNP'] = df['SNP'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['A1'] = df['A1'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['A2'] = df['A2'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['MAF'] = df['MAF'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['Effect'] = df['Effect'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['SE'] = df['SE'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['Pval'] = df['Pval'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "\n",
    "# Reorder according to CHR and POS\n",
    "df.sort_values(by=['CHR', 'POS'], ascending=True, inplace=True)\n",
    "\n",
    "\n",
    "# Re-code 'NA' to .\n",
    "df.replace(to_replace=['NA'],\n",
    "            value='.',\n",
    "            inplace=True\n",
    "            )\n",
    "\n",
    "# Save\n",
    "save_filename = \"pQC.DIAMANTE-EUR.sumstat.txt\"\n",
    "df.to_csv(save_filename, sep='\\t', index=False)\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Type II Diabetes (T2D) / Cohort: MVP\n",
    "\n",
    "- File name: T2D_MVP_only_EUR_dbGaP.txt\n",
    "- Whitespace-separated\n",
    "- Columns: RSID CHR POS NEA EA EAF BETA SE P N MACH_R2 CALLRATE HWE OR\n",
    "- GRCh37\n",
    "\n",
    "README:\\\n",
    "RSID\tReference SNP identifier. rs number (RSID) assigned by dbSNP\\\n",
    "CHR\tchromosomal number: 1, 2 . .,22\\\n",
    "POS\tLocus position on a chromosome (hg19)\\\n",
    "NEA\tNon-effect allele\\\n",
    "EA\tEffect allele\\\n",
    "EAF\tEffect allele frequency across all samples\\\n",
    "BETA\tRegression coefficient (이건 log(OR).)\\\n",
    "SE\tStandard error of estimated effect\\\n",
    "P\tP value based on test statistic\\\n",
    "N\tNumber of subjects participated in data analysis\\\n",
    "MACH_R2\timputation quality\\\n",
    "CALLRATE\tcall rate for genotyped SNPs\\\n",
    "HWE\tHWE p-value\\\n",
    "OR\tOdds ratio\\\n",
    "\n",
    "\n",
    "QC performed:\n",
    "- Drop unneccessary columns: HWE, CALLRATE\n",
    "- Rename columns\n",
    "- Code missing data as NA\n",
    "- MAF check\n",
    "- Rename columns\n",
    "- Reorder columns\n",
    "- Assign appropriate data type except NA\n",
    "- Re-code 'NA' to ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "from math import log\n",
    "\n",
    "file = 'T2D_MVP_only_EUR_dbGaP.txt'\n",
    "df = pd.read_csv(file, sep=' ', index_col=False)\n",
    "\n",
    "\n",
    "# Drop unneccessary columns: chrposID\n",
    "df.drop(columns=['HWE', 'CALLRATE'], inplace=True)\n",
    "\n",
    "# Rename columns\n",
    "df.rename(columns={'RSID': 'SNP', 'NEA':'A2', 'EA':'A1', 'BETA':'Effect', 'P':'Pval', 'MACH_R2':'INFO'}, inplace=True)\n",
    "\n",
    "# Code missing data as NA\n",
    "df['CHR'] = df['CHR'].apply(lambda x: 'NA' if is_na(x) else str(int(x)))\n",
    "df['POS'] = df['POS'].apply(lambda x: 'NA' if is_na(x) else str(int(x)))\n",
    "df['SNP'] = df['SNP'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['A1'] = df['A1'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['A2'] = df['A2'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['Pval'] = df['Pval'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['Effect'] = df['Effect'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['SE'] = df['SE'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['EAF'] = df['EAF'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['N'] = df['N'].apply(lambda x: 'NA' if is_na(x) else str(int(x)))\n",
    "df['OR'] = df['OR'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['INFO'] = df['INFO'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "\n",
    "# MAF check\n",
    "df['MAF check'] = df['EAF'].apply(lambda x: 'NA' if x == 'NA' else True if float(x) <= 0.5 else False)\n",
    "df['MAF'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['EAF'] if row['MAF check'] else str(1 - float(row['EAF'])), axis=1)\n",
    "df['A1_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['A1'] if row['MAF check'] else row['A2'], axis=1)\n",
    "df['A2_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['A2'] if row['MAF check'] else row['A1'], axis=1)\n",
    "df['Effect_new'] =  df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['Effect'] if row['MAF check'] else str(-1 * float(row['Effect'])), axis=1)\n",
    "df.drop(columns=['EAF', 'MAF check', 'A1', 'A2', 'Effect'], inplace=True)\n",
    "\n",
    "df.rename(columns={'A1_new':'A1', 'A2_new':'A2', 'Effect_new':'Effect'}, inplace=True)\n",
    "\n",
    "\n",
    "# Reorder columns\n",
    "df = df[['CHR', 'POS', 'SNP', 'A1', 'A2', 'MAF', 'Effect', 'OR', 'SE', 'Pval', 'N', 'INFO']]\n",
    "\n",
    "# Assign appropriate data type except NA\n",
    "df['CHR'] = df['CHR'].apply(lambda x: 'NA' if x == 'NA' else int(x))\n",
    "df['POS'] = df['POS'].apply(lambda x: 'NA' if x == 'NA' else int(x))\n",
    "df['SNP'] = df['SNP'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['A1'] = df['A1'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['A2'] = df['A2'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['MAF'] = df['MAF'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['Effect'] = df['Effect'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['OR'] = df['OR'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['SE'] = df['SE'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['Pval'] = df['Pval'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['N'] = df['N'].apply(lambda x: 'NA' if x == 'NA' else int(x))\n",
    "df['INFO'] = df['INFO'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "\n",
    "# Reorder according to CHR and POS\n",
    "df.sort_values(by=['CHR', 'POS'], ascending=True, inplace=True)\n",
    "\n",
    "\n",
    "# Re-code 'NA' to .\n",
    "df.replace(to_replace=['NA'],\n",
    "            value='.',\n",
    "            inplace=True\n",
    "            )\n",
    "\n",
    "# Save\n",
    "save_filename = \"pQC.T2D_MVP_only_EUR_dbGaP.txt\"\n",
    "df.to_csv(save_filename, sep='\\t', index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Type II Diabetes (T2D) / Cohort: [METAL] FinnGen release 7, Mahajan 2022, MVP\n",
    "\n",
    "- File name: ./meta_analysis/T2D_FinngenR7_Mahajan2022_MVP/meta.T2D.FinngenR7.Mahajan2022.MVP.1.tbl\n",
    "- tab-separated\n",
    "- Columns: MarkerName      Allele1 Allele2 Freq1   FreqSE  Effect  StdErr  P-value Direction       HetISq  HetChiSq        HetDf   HetPVal\n",
    "- GRCh37\n",
    "\n",
    "README:\\\n",
    "\n",
    "\n",
    "QC performed:\n",
    "- Remove unneccesary column: FreqSE, Direction, HetISq, HetChiSq, HetDf, HetPVal\n",
    "- Code missing data as NA\n",
    "- CHR, POS matching from 1kG\n",
    "- MAF check\n",
    "- Rename columns\n",
    "- Reorder columns\n",
    "- Assign appropriate data type except NA\n",
    "- Re-code 'NA' to ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "from math import log, exp\n",
    "\n",
    "file = \"./meta_analysis/T2D_FinngenR7_Mahajan2022_MVP/meta.T2D.FinngenR7.Mahajan2022.MVP.1.tbl\"\n",
    "df = pd.read_csv(file, sep='\\t', index_col=False)\n",
    "\n",
    "# Remove unneccesary columns\n",
    "df.drop(columns=['FreqSE', 'Direction', 'HetISq', 'HetChiSq', 'HetDf', 'HetPVal'], inplace=True)\n",
    "\n",
    "# Code missing data as NA\n",
    "df['MarkerName'] = df['MarkerName'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['Allele1'] = df['Allele1'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['Allele2'] = df['Allele2'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['Freq1'] = df['Freq1'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['Effect'] = df['Effect'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['StdErr'] = df['StdErr'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['P-value'] = df['P-value'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "\n",
    "# CHR and POS from 1kG\n",
    "df = df.merge(var_match_ref, left_on='MarkerName', right_on='rsID', how='left')\n",
    "\n",
    "def check_chrpos_na(x):\n",
    "    try:\n",
    "        if is_floatNan(x):\n",
    "            return True\n",
    "        return False\n",
    "    except:\n",
    "        if is_strna(x):\n",
    "            return True\n",
    "        return False\n",
    "\n",
    "df['CHRPOS'] = df['CHRPOS'].apply(lambda x: 'NA' if check_chrpos_na(x) else str(x))\n",
    "df['CHR'] = df['CHRPOS'].apply(lambda uniqid: 'NA' if uniqid == 'NA' else str(uniqid.split(sep=':')[0]))\n",
    "df['POS'] = df['CHRPOS'].apply(lambda uniqid: 'NA' if uniqid == 'NA' else str(uniqid.split(sep=':')[1]))\n",
    "df.drop(columns=['CHRPOS', 'rsID'], inplace=True)\n",
    "\n",
    "# MAF calculation and check\n",
    "df['MAF check'] = df['Freq1'].apply(lambda x: 'NA' if x == 'NA' else True if float(x) <= 0.5 else False)\n",
    "df['MAF_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['Freq1'] if row['MAF check'] else str(1 - float(row['Freq1'])), axis=1)\n",
    "df['A1_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['Allele1'] if row['MAF check'] else row['Allele2'], axis=1)\n",
    "df['A2_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['Allele2'] if row['MAF check'] else row['Allele1'], axis=1)\n",
    "df['Effect_new'] =  df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['Effect'] if row['MAF check'] else str(-1 * float(row['Effect'])), axis=1)\n",
    "df.drop(columns=['Freq1', 'MAF check', 'Allele1', 'Allele2', 'Effect'], inplace=True)\n",
    "\n",
    "# Rename columns\n",
    "df.rename(columns={'MarkerName':'SNP', 'A1_new':'A1', 'A2_new':'A2', 'Effect_new':'Effect', 'MAF_new':'MAF', 'StdErr':'SE', 'P-value':'Pval'}, inplace=True)\n",
    "\n",
    "# Reorder columns\n",
    "df = df[['CHR', 'POS', 'SNP', 'A1', 'A2', 'MAF', 'Effect', 'SE', 'Pval']]\n",
    "\n",
    "# Assign appropriate data type except NA\n",
    "df['CHR'] = df['CHR'].apply(lambda x: 'NA' if x == 'NA' else int(x))\n",
    "df['POS'] = df['POS'].apply(lambda x: 'NA' if x == 'NA' else int(x))\n",
    "df['SNP'] = df['SNP'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['A1'] = df['A1'].apply(lambda x: 'NA' if x == 'NA' else str(x).upper())\n",
    "df['A2'] = df['A2'].apply(lambda x: 'NA' if x == 'NA' else str(x).upper())\n",
    "df['MAF'] = df['MAF'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['Effect'] = df['Effect'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['SE'] = df['SE'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['Pval'] = df['Pval'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "\n",
    "# Reorder according to CHR and POS\n",
    "df.sort_values(by=['CHR', 'POS'], ascending=True, inplace=True)\n",
    "\n",
    "\n",
    "# Re-code 'NA' to .\n",
    "df.replace(to_replace=['NA'],\n",
    "            value='.',\n",
    "            inplace=True\n",
    "            )\n",
    "\n",
    "# Save\n",
    "save_filename = os.path.join(dir, \"meta.T2D.FinngenR7.Mahajan2022.MVP.1.tbl.txt\")\n",
    "df.to_csv(save_filename, sep='\\t', index=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Type II Diabetes (T2D) / Cohort: [METAL] FinnGen release 7, MVP\n",
    "\n",
    "- File name: ./meta_analysis/T2D_FinngenR7_MVP/meta.T2D.FinngenR7.MVP.1.tbl\n",
    "- tab-separated\n",
    "- Columns: MarkerName      Allele1 Allele2 Freq1   FreqSE  Effect  StdErr  P-value Direction       HetISq  HetChiSq        HetDf   HetPVal\n",
    "- GRCh37\n",
    "\n",
    "README:\\\n",
    "\n",
    "\n",
    "QC performed:\n",
    "- Remove unneccesary column: FreqSE, Direction, HetISq, HetChiSq, HetDf, HetPVal\n",
    "- Code missing data as NA\n",
    "- CHR, POS matching from 1kG\n",
    "- MAF check\n",
    "- Rename columns\n",
    "- Reorder columns\n",
    "- Assign appropriate data type except NA\n",
    "- Re-code 'NA' to ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "from math import log, exp\n",
    "\n",
    "file = \"./meta_analysis/T2D_FinngenR7_MVP/meta.T2D.FinngenR7.MVP.1.tbl\"\n",
    "df = pd.read_csv(file, sep='\\t', index_col=False)\n",
    "\n",
    "# Remove unneccesary columns\n",
    "df.drop(columns=['FreqSE', 'Direction', 'HetISq', 'HetChiSq', 'HetDf', 'HetPVal'], inplace=True)\n",
    "\n",
    "# Code missing data as NA\n",
    "df['MarkerName'] = df['MarkerName'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['Allele1'] = df['Allele1'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['Allele2'] = df['Allele2'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['Freq1'] = df['Freq1'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['Effect'] = df['Effect'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['StdErr'] = df['StdErr'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['P-value'] = df['P-value'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "\n",
    "# CHR and POS from 1kG\n",
    "df = df.merge(var_match_ref, left_on='MarkerName', right_on='rsID', how='left')\n",
    "\n",
    "def check_chrpos_na(x):\n",
    "    try:\n",
    "        if is_floatNan(x):\n",
    "            return True\n",
    "        return False\n",
    "    except:\n",
    "        if is_strna(x):\n",
    "            return True\n",
    "        return False\n",
    "\n",
    "df['CHRPOS'] = df['CHRPOS'].apply(lambda x: 'NA' if check_chrpos_na(x) else str(x))\n",
    "df['CHR'] = df['CHRPOS'].apply(lambda uniqid: 'NA' if uniqid == 'NA' else str(uniqid.split(sep=':')[0]))\n",
    "df['POS'] = df['CHRPOS'].apply(lambda uniqid: 'NA' if uniqid == 'NA' else str(uniqid.split(sep=':')[1]))\n",
    "df.drop(columns=['CHRPOS', 'rsID'], inplace=True)\n",
    "\n",
    "# MAF calculation and check\n",
    "df['MAF check'] = df['Freq1'].apply(lambda x: 'NA' if x == 'NA' else True if float(x) <= 0.5 else False)\n",
    "df['MAF_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['Freq1'] if row['MAF check'] else str(1 - float(row['Freq1'])), axis=1)\n",
    "df['A1_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['Allele1'] if row['MAF check'] else row['Allele2'], axis=1)\n",
    "df['A2_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['Allele2'] if row['MAF check'] else row['Allele1'], axis=1)\n",
    "df['Effect_new'] =  df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['Effect'] if row['MAF check'] else str(-1 * float(row['Effect'])), axis=1)\n",
    "df.drop(columns=['Freq1', 'MAF check', 'Allele1', 'Allele2', 'Effect'], inplace=True)\n",
    "\n",
    "# Rename columns\n",
    "df.rename(columns={'MarkerName':'SNP', 'A1_new':'A1', 'A2_new':'A2', 'Effect_new':'Effect', 'MAF_new':'MAF', 'StdErr':'SE', 'P-value':'Pval'}, inplace=True)\n",
    "\n",
    "# Reorder columns\n",
    "df = df[['CHR', 'POS', 'SNP', 'A1', 'A2', 'MAF', 'Effect', 'SE', 'Pval']]\n",
    "\n",
    "# Assign appropriate data type except NA\n",
    "df['CHR'] = df['CHR'].apply(lambda x: 'NA' if x == 'NA' else int(x))\n",
    "df['POS'] = df['POS'].apply(lambda x: 'NA' if x == 'NA' else int(x))\n",
    "df['SNP'] = df['SNP'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['A1'] = df['A1'].apply(lambda x: 'NA' if x == 'NA' else str(x).upper())\n",
    "df['A2'] = df['A2'].apply(lambda x: 'NA' if x == 'NA' else str(x).upper())\n",
    "df['MAF'] = df['MAF'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['Effect'] = df['Effect'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['SE'] = df['SE'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['Pval'] = df['Pval'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "\n",
    "# Reorder according to CHR and POS\n",
    "df.sort_values(by=['CHR', 'POS'], ascending=True, inplace=True)\n",
    "\n",
    "\n",
    "# Re-code 'NA' to .\n",
    "df.replace(to_replace=['NA'],\n",
    "            value='.',\n",
    "            inplace=True\n",
    "            )\n",
    "\n",
    "# Save\n",
    "save_filename = \"meta.T2D.FinngenR7.MVP.1.tbl.txt\"\n",
    "df.to_csv(save_filename, sep='\\t', index=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Waist circumference (WC)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Waist circumference / Cohort: UKB\n",
    "\n",
    "- File name: f.48.0.0_res.EUR.sumstats.MACfilt.txt\n",
    "- tab-separated\n",
    "- Columns: SNP     CHR     BP      A1      TEST    NMISS   BETA    SE      L95     U95     STAT    P       A2      MAF     NCHROBS SNPID_UKB       A1_UKB  A2_UKB  INFO_UKB        MAF_UKB\n",
    "- GRCh37\n",
    "\n",
    "README:\\\n",
    "SNP: unique ID of the SNP consists of chromosome, position and alphabetically ordered alleles\\\n",
    "CHR: chromosome\\\n",
    "BP: base pair position on GRCh37\\\n",
    "A1: effect allele\\\n",
    "TEST: Type of test (ADD for all files)\\\n",
    "NMISS: Number of non-missing genotypes\\\n",
    "BETA/OR: Regression coefficient or odds ratio\\\n",
    "SE: Standard error (for binary traits, SE in logOR scale)\\\n",
    "L95: Lower bound on confidence interval for CMH odds ratio\\\n",
    "U95: Upper bound on confidence interval for CMH odds ratio\\\n",
    "STAT: Coefficient t-statistics\\\n",
    "P: P-value\\\n",
    "A2: non effect allele\\\n",
    "MAF: Minor allele frequency\\\n",
    "NCHROBS: Number of allele observation\\\n",
    "SNPID_UKB: rsID provided by UK Biobank\\\n",
    "A1_UKB: A1 allele in UK Biobank\\\n",
    "A2_UKB: A2 allele in UK Biobank\\\n",
    "INFO_UKB: Info score provided by UK Biobank\\\n",
    "MAF_UKB: MAF of entire UK Boiobank samples\\\n",
    "\n",
    "QC performed:\n",
    "- Remove unneccesary column: TEST, NMISS, L95, U95, STAT, NCHROBS, A1_UKB, A2_UKB, MAF_UKB\n",
    "- Code missing data as NA\n",
    "- Rename columns\n",
    "- Reorder columns\n",
    "- Assign appropriate data type except NA\n",
    "- Re-code 'NA' to ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "file = \"f.48.0.0_res.EUR.sumstats.MACfilt.txt\"\n",
    "df = pd.read_csv(file, sep='\\t', index_col=False)\n",
    "\n",
    "# Remove unneccesary columns\n",
    "df.drop(columns=['SNP', 'TEST', 'NMISS', 'L95', 'U95', 'STAT', 'NCHROBS', 'A1_UKB', 'A2_UKB', 'MAF_UKB'], inplace=True)\n",
    "\n",
    "# Code missing data as NA\n",
    "df['SNPID_UKB'] = df['SNPID_UKB'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['CHR'] = df['CHR'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['BP'] = df['BP'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['A1'] = df['A1'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['A2'] = df['A2'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['MAF'] = df['MAF'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['BETA'] = df['BETA'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['SE'] = df['SE'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['P'] = df['P'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "\n",
    "# Rename columns\n",
    "df.rename(columns={'SNPID_UKB':'SNP', 'BP':'POS', 'BETA':'Effect', 'P':'Pval'}, inplace=True)\n",
    "\n",
    "# Reorder columns\n",
    "df = df[['CHR', 'POS', 'SNP', 'A1', 'A2', 'MAF', 'Effect', 'SE', 'Pval']]\n",
    "\n",
    "# Assign appropriate data type except NA\n",
    "df['CHR'] = df['CHR'].apply(lambda x: 'NA' if x == 'NA' else int(x))\n",
    "df['POS'] = df['POS'].apply(lambda x: 'NA' if x == 'NA' else int(x))\n",
    "df['SNP'] = df['SNP'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['A1'] = df['A1'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['A2'] = df['A2'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['MAF'] = df['MAF'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['Effect'] = df['Effect'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['SE'] = df['SE'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['Pval'] = df['Pval'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "\n",
    "# Reorder according to CHR and POS\n",
    "df.sort_values(by=['CHR', 'POS'], ascending=True, inplace=True)\n",
    "\n",
    "\n",
    "# Re-code 'NA' to .\n",
    "df.replace(to_replace=['NA'],\n",
    "            value='.',\n",
    "            inplace=True\n",
    "            )\n",
    "\n",
    "# Save\n",
    "save_filename = \"pQC.f.48.0.0_res.EUR.sumstats.MACfilt.txt\"\n",
    "df.to_csv(save_filename, sep='\\t', index=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Waist circumference / Cohort: GIANT 2015\n",
    "\n",
    "- File name: GIANT_2015_WC_COMBINED_EUR.txt\n",
    "- tab-separated\n",
    "- Columns: \n",
    "- GRCh37\n",
    "\n",
    "README:\\\n",
    "MarkerName: The dbSNP name of the genetic marker \\\n",
    "Allele1: The first allele (hg19 + strand). Where the regression coefficients (betas) are provided, the first allele is the effect allele. Where betas are not provided (typically the 2010 data), the first allele is the trait-increasing allele. \\\n",
    "Allele2: The second allele (hg19 + strand) \\\n",
    "Freq.Allele1.HapMapCEU: The allele frequency of Allele1 in the HapMap CEU population \\\n",
    "b: beta \\\n",
    "SE: standard error \\\n",
    "p: p-value after meta-analysis using regression coefficients (beta and standard error), and after correction for inflation of test statistics using genomic control both at the individual study level and again after meta-analysis \\\n",
    "N: Number of observations \\\n",
    "\n",
    "\n",
    "QC performed:\n",
    "- Remove binary rows\n",
    "- Code missing data as NA\n",
    "- Get CHR and POS\n",
    "- Check MAF\n",
    "- Rename columns\n",
    "- Reorder columns\n",
    "- Assign appropriate data type except NA\n",
    "- Re-code 'NA' to ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "file = \"GIANT_2015_WC_COMBINED_EUR.txt\"\n",
    "with open(file, 'rb') as f:\n",
    "    lines = f.readlines()\n",
    "\n",
    "data = []\n",
    "for idx, line in enumerate(lines):\n",
    "    if idx <= 7:\n",
    "        continue\n",
    "    data.append(lines[idx].decode(\"utf-8\").rstrip().split(sep='\\t'))\n",
    "\n",
    "df = pd.DataFrame(data, columns=['SNP', 'A1', 'A2', 'MAF', 'Effect', 'SE', 'Pval', 'N'])\n",
    "\n",
    "# Code missing data as NA\n",
    "df['SNP'] = df['SNP'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['A1'] = df['A1'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['A2'] = df['A2'].apply(lambda x: 'NA' if is_na(x) else str(x))\n",
    "df['MAF'] = df['MAF'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['Effect'] = df['Effect'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['SE'] = df['SE'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['Pval'] = df['Pval'].apply(lambda x: 'NA' if is_na(x) else str(float(x)))\n",
    "df['N'] = df['N'].apply(lambda x: 'NA' if is_na(x) else str(int(x)))\n",
    "\n",
    "# CHR and POS from 1kG\n",
    "df = df.merge(var_match_ref, left_on='SNP', right_on='rsID', how='left')\n",
    "\n",
    "def check_chrpos_na(x):\n",
    "    try:\n",
    "        if is_floatNan(x):\n",
    "            return True\n",
    "        return False\n",
    "    except:\n",
    "        if is_strna(x):\n",
    "            return True\n",
    "        return False\n",
    "\n",
    "df['CHRPOS'] = df['CHRPOS'].apply(lambda x: 'NA' if check_chrpos_na(x) else str(x))\n",
    "df['CHR'] = df['CHRPOS'].apply(lambda uniqid: 'NA' if uniqid == 'NA' else str(uniqid.split(sep=':')[0]))\n",
    "df['POS'] = df['CHRPOS'].apply(lambda uniqid: 'NA' if uniqid == 'NA' else str(uniqid.split(sep=':')[1]))\n",
    "df.drop(columns=['CHRPOS', 'rsID'], inplace=True)\n",
    "\n",
    "# MAF calculation and check\n",
    "df['MAF check'] = df['MAF'].apply(lambda x: 'NA' if x == 'NA' else True if float(x) <= 0.5 else False)\n",
    "df['MAF_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['MAF'] if row['MAF check'] else str(1 - float(row['MAF'])), axis=1)\n",
    "df['A1_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['A1'] if row['MAF check'] else row['A2'], axis=1)\n",
    "df['A2_new'] = df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['A2'] if row['MAF check'] else row['A1'], axis=1)\n",
    "df['Effect_new'] =  df.apply(lambda row: 'NA' if row['MAF check'] == 'NA' else row['Effect'] if row['MAF check'] else str(-1 * float(row['Effect'])), axis=1)\n",
    "df.drop(columns=['MAF', 'MAF check', 'A1', 'A2', 'Effect'], inplace=True)\n",
    "\n",
    "\n",
    "# Rename columns\n",
    "df.rename(columns={'MAF_new': 'MAF', 'A1_new':'A1', 'A2_new':'A2', 'Effect_new':'Effect'}, inplace=True)\n",
    "\n",
    "# Reorder columns\n",
    "df = df[['CHR', 'POS', 'SNP', 'A1', 'A2', 'MAF', 'Effect', 'SE', 'Pval', 'N']]\n",
    "\n",
    "# Assign appropriate data type except NA\n",
    "df['CHR'] = df['CHR'].apply(lambda x: 'NA' if x == 'NA' else int(x))\n",
    "df['POS'] = df['POS'].apply(lambda x: 'NA' if x == 'NA' else int(x))\n",
    "df['SNP'] = df['SNP'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['A1'] = df['A1'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['A2'] = df['A2'].apply(lambda x: 'NA' if x == 'NA' else str(x))\n",
    "df['MAF'] = df['MAF'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['Effect'] = df['Effect'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['SE'] = df['SE'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['Pval'] = df['Pval'].apply(lambda x: 'NA' if x == 'NA' else float(x))\n",
    "df['N'] = df['N'].apply(lambda x: 'NA' if x == 'NA' else int(x))\n",
    "\n",
    "# Reorder according to CHR and POS\n",
    "df.sort_values(by=['CHR', 'POS'], ascending=True, inplace=True)\n",
    "\n",
    "\n",
    "# Re-code 'NA' to .\n",
    "df.replace(to_replace=['NA'],\n",
    "            value='.',\n",
    "            inplace=True\n",
    "            )\n",
    "\n",
    "# Save\n",
    "save_filename = \"pQC.GIANT_2015_WC_COMBINED_EUR.txt\"\n",
    "df.to_csv(save_filename, sep='\\t', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.7.13 ('base')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.13"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "4064372752f144f7b8f9f770c681ae1da5b167ff29634c4a414d8cf16e139af0"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
